File: python/day15.py
def eval(seed, stop_k):
    positions = {}
    next_num = 0
    k = 1
    while True:
        num = seed[k - 1] if k <= len(seed) else next_num
        if k == stop_k:
            return num

        if num in positions:
            next_num = k - positions[num]
        else:
            next_num = 0

        positions[num] = k
        k += 1


def solution():
    seed = [int(s) for s in open('../data/15.txt', 'r').read().split(',')]

    print(f'Answer 1: {eval(seed, 2020)}')
    print(f'Answer 2: {eval(seed, 30000000)}')

File: rust/src/day15.rs
use crate::common;
use std::collections::HashMap;

fn eval(seed: &Vec<u64>, stop_k: u64) -> u64 {
  let mut positions = HashMap::new();
  let mut next_num = 0;
  let mut k: usize = 1;
  loop {
    let num = if k <= seed.len() {
      seed[k - 1]
    } else {
      next_num
    };
    if k as u64 == stop_k {
      return num;
    }
    match positions.get(&num) {
      Some(&x) => next_num = (k as u64) - x,
      _ => next_num = 0,
    };
    positions.insert(num, k as u64);
    k += 1;
  }
}

pub(crate) fn solution() {
  let seed = std::fs::read_to_string(&common::data_file(15))
    .unwrap()
    .split(",")
    .map(|x| x.parse::<u64>().unwrap())
    .collect();
  println!("Answer 1: {}", eval(&seed, 2020));
  println!("Answer 2: {}", eval(&seed, 30000000));
}
